### DISCUSSION TITLE
How do the gPCE coefficients in Dakota's output relate to equations in Theory Manual?

### DISCUSSION THREAD

**Discussion by sitadrost** [snl-dakota/dakota#11] [Q&A] [answered]
Dear all,

After computing a polynomial chaos expansion on a black box simulation model, Dakota's output contains lists/tables with coefficients for each response variable, for example (1st order expansion, Askey, model with 9 variables):
```
Coefficients of Polynomial Chaos Expansion for R1:
        coefficient   u1   u2   u3   u4   u5   u6   u7   u8   u9
        ----------- ---- ---- ---- ---- ---- ---- ---- ---- ----
   5.4055994585e-01  He0  He0  He0  He0  He0  He0  He0  He0   P0
   3.4964094113e-03  He1  He0  He0  He0  He0  He0  He0  He0   P0
   3.1452953521e-04  He0  He1  He0  He0  He0  He0  He0  He0   P0
   4.1718308256e-03  He0  He0  He1  He0  He0  He0  He0  He0   P0
  -2.8956633942e-02  He0  He0  He0  He1  He0  He0  He0  He0   P0
   1.2889486004e-02  He0  He0  He0  He0  He1  He0  He0  He0   P0
   1.9541049530e-03  He0  He0  He0  He0  He0  He1  He0  He0   P0
  -1.0685641734e-02  He0  He0  He0  He0  He0  He0  He1  He0   P0
  -1.3524728925e-02  He0  He0  He0  He0  He0  He0  He0  He1   P0
   9.7442036768e-02  He0  He0  He0  He0  He0  He0  He0  He0   P1
```
where `u1` to `u8` are normally distributed and `u9` has uniform distribution.
Based on Table 3.1 in the 6.16 Theory Manual (I haven't upgraded to 6.17 yet), I'm guessing that `He0` means 0th-order Hermite polynomial, `P1` 1st-order Legendre polynomial, etc. Is that correct?
I was wondering, when using these coefficients to construct the gPCE for this response variable, following eq (3.17)/(3.20), should I scale my variables to standard normal/uniform distributions? And is the output scaled as well?

Many thanks,
Sita

**Comment by mseldre** [comment, marked as answer]
You are correct about the short-hand for Hermite (He) and Legendre (P) along with the basis order.  By default, the polynomials are not normalized to unit variance, although this is a specification option. You do not need to scale your input variables in either case -- Dakota takes care of that for you (see also section 3.5 in the theory manual version you're using for options).

The output is not scaled -- it is projected as is against the basis polynomials that are employed.

You may find it instructive to start from a response QoI that is a known polynomial of low order, something like Rosenbrock's function in two dimensions or similar.  You can take the coefficients Dakota computes, multiply them by the low order Hermite or Legendre polynomials, and exactly reproduce your starting polynomial (assuming the order of the expansion is sufficient). 

For example, 2D Rosenbrock in a Hermite basis:
   **4.0200000000e+02  He0  He0**
  **-2.0000000000e+00  He1  He0**
   **6.0100000000e+02  He2  He0**
   2.3684757859e-15  He3  He0
   **1.0000000000e+02  He4  He0
  -2.0000000000e+02  He0  He1**
   0.0000000000e+00  He1  He1
  **-2.0000000000e+02  He2  He1**
  -1.8947806287e-14  He3  He1
   3.7895612574e-14  He4  He1
   **1.0000000000e+02  He0  He2**
  -3.5527136788e-15  He1  He2
  -7.1054273576e-14  He2  He2
   2.3684757859e-15  He3  He2
   1.1842378929e-15  He4  He2

and in a Legendre basis:
   **4.5566666667e+02   P0   P0
  -4.0000000000e+00   P1   P0
   9.1695238095e+02   P2   P0**
  -5.5955240441e-14   P3   P0
   **3.6571428571e+02   P4   P0
  -5.3333333333e+02   P0   P1**
  -4.7961634664e-14   P1   P1
  **-1.0666666667e+03   P2   P1**
  -8.3932860662e-14   P3   P1
   1.1151080059e-12   P4   P1
   **2.6666666667e+02   P0   P2**
   6.6613381478e-14   P1   P2
   2.2204460493e-13   P2   P2
   7.7715611724e-15   P3   P2
  -4.9960036108e-13   P4   P2''

**Reply by sitadrost** [reply]
Thanks a lot for your explanation! That's a good one, to start from a known response, I should have started with that of course, but I was super curious to see my own results.

**Comment by mseldre** [comment]
Thanks for pointing out the loss of some valuable docs in the new manual set — we definitely need to fix that.

The numerical techniques that are used are as described in that paragraph.  The source code that is used for this resides in dakota/packages/pecos/src/NumericGenOrthogPolynomial.cpp within the function NumericGenOrthogPolynomial::solve_eigenproblem(), where you can see a few special cases (e.g. LOGNORMAL) that have short-cuts alongside a default implementation that cites the discretized Stieltjes algorithm (W. Gautschi, SIAM J. Sci. Stat. Comput., Vol. 3, No. 3, Sept. 1982) for computing inner products following by solution of a symmetric tridiagonal eigenvalue problem using LAPACK.

I realize that reading source code may not be the most helpful, but there are a number of comments that enumerate the steps alongside the code.  So hopefully this gives you a better idea.

Note that the “polyCoeffs” are the coefficients of the orthogonal polynomials and are output for debug-level verbosity, so you can reconstruct the univariate basis polynomials from these if you wish.

Best regards,
—Mike

--

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Michael S. Eldred, Ph.D.

Sandia National Laboratories

P.O. Box 5800, Mail Stop 1318

Albuquerque, NM 87185-1318

(505)844-6479, FAX:(505)284-2518

***@***.******@***.***>

https://www.sandia.gov/ccr/staff/michael-s-eldred/

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

From: sitadrost ***@***.***>
Date: Tuesday, September 30, 2025 at 2:45 AM
To: snl-dakota/dakota ***@***.***>
Cc: Eldred, Michael S. ***@***.***>, Comment ***@***.***>
Subject: [EXTERNAL] Re: [snl-dakota/dakota] How do the gPCE coefficients in Dakota's output relate to equations in Theory Manual? (Discussion #11)


Following up on my earlier question, I was wondering how these numerically generated orthogonal polynomials work:

Coefficients of Polynomial Chaos Expansion for R1:
        coefficient   u1   u2   u3   u4   u5   u6   u7
        ----------- ---- ---- ---- ---- ---- ---- ----
   1.7642338446e+00 Num0 Num0 Num0 Num0 Num0 Num0   P0
   7.8350046724e-04 Num1 Num0 Num0 Num0 Num0 Num0   P0
   4.5430569440e-03 Num0 Num1 Num0 Num0 Num0 Num0   P0
   8.4910279301e-01 Num0 Num0 Num1 Num0 Num0 Num0   P0
  -9.0134378903e-06 Num0 Num0 Num0 Num1 Num0 Num0   P0
   2.6503945379e+00 Num0 Num0 Num0 Num0 Num1 Num0   P0
  -1.3398933210e-02 Num0 Num0 Num0 Num0 Num0 Num1   P0
  -1.1126509255e-01 Num0 Num0 Num0 Num0 Num0 Num0   P1


u1 to u6 are bounded normal variables here, u7 has uniform distribution.

According to the User Manual:

Extended (default if no option is selected): The Extended option avoids the use of any nonlinear variable transformations by augmenting the Askey approach with numerically-generated orthogonal polynomials for non-Askey probability density functions. Extended polynomial selections with numerically-generated polynomials that are orthogonal to the prescribed probability density functions replace each of the sub-optimal Askey basis selections for bounded normal, lognormal, bounded lognormal, loguniform, triangular, gumbel, frechet, weibull, and bin-based histogram.

My old version 6.16.0 Theory Manual gives some additional information:

An alternative is to numerically generate the orthogonal polynomials (using Gauss-Wigert [120], discretized Stieltjes [59], Chebyshev [59], or Gramm-Schmidt [138] approaches) and then compute their Gauss points and weights (using the Golub-Welsch [69] tridiagonal eigensolution). These solutions are optimal for given random variable sets having arbitrary probability density functions and eliminate the need to induce additional nonlinearity through variable transformations, but performing this process for general joint density functions with correlation is a topic of ongoing research (refer to Section 3.5 for additional details).

Where can I find more information on these numerically-generated orthogonal polynomials? Would it be possible for me to reconstruct the gPCE using the coefficients provided by Dakota?

**Comment by mseldre** [comment]
Correction: output of the numerically-generated polynomial coefficients and other quantities requires recompiling with the line

    #define DEBUG

uncommented, not simply specifying “output debug” in the Dakota input file.

--Mike

--

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Michael S. Eldred, Ph.D.

Sandia National Laboratories

P.O. Box 5800, Mail Stop 1318

Albuquerque, NM 87185-1318

(505)844-6479, FAX:(505)284-2518

***@***.******@***.***>

https://www.sandia.gov/ccr/staff/michael-s-eldred/

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

From: Eldred, Michael S. ***@***.***>
Date: Wednesday, October 1, 2025 at 5:40 PM
To: snl-dakota/dakota ***@***.***>, snl-dakota/dakota ***@***.***>
Cc: Comment ***@***.***>, Stephens, Adam ***@***.***>
Subject: Re: [EXTERNAL] Re: [snl-dakota/dakota] How do the gPCE coefficients in Dakota's output relate to equations in Theory Manual? (Discussion #11)

Thanks for pointing out the loss of some valuable docs in the new manual set — we definitely need to fix that.

The numerical techniques that are used are as described in that paragraph.  The source code that is used for this resides in dakota/packages/pecos/src/NumericGenOrthogPolynomial.cpp within the function NumericGenOrthogPolynomial::solve_eigenproblem(), where you can see a few special cases (e.g. LOGNORMAL) that have short-cuts alongside a default implementation that cites the discretized Stieltjes algorithm (W. Gautschi, SIAM J. Sci. Stat. Comput., Vol. 3, No. 3, Sept. 1982) for computing inner products following by solution of a symmetric tridiagonal eigenvalue problem using LAPACK.

I realize that reading source code may not be the most helpful, but there are a number of comments that enumerate the steps alongside the code.  So hopefully this gives you a better idea.

Note that the “polyCoeffs” are the coefficients of the orthogonal polynomials and are output for debug-level verbosity, so you can reconstruct the univariate basis polynomials from these if you wish.

Best regards,
—Mike

--

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Michael S. Eldred, Ph.D.

Sandia National Laboratories

P.O. Box 5800, Mail Stop 1318

Albuquerque, NM 87185-1318

(505)844-6479, FAX:(505)284-2518

***@***.******@***.***>

https://www.sandia.gov/ccr/staff/michael-s-eldred/

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

From: sitadrost ***@***.***>
Date: Tuesday, September 30, 2025 at 2:45 AM
To: snl-dakota/dakota ***@***.***>
Cc: Eldred, Michael S. ***@***.***>, Comment ***@***.***>
Subject: [EXTERNAL] Re: [snl-dakota/dakota] How do the gPCE coefficients in Dakota's output relate to equations in Theory Manual? (Discussion #11)


Following up on my earlier question, I was wondering how these numerically generated orthogonal polynomials work:

Coefficients of Polynomial Chaos Expansion for R1:
        coefficient   u1   u2   u3   u4   u5   u6   u7
        ----------- ---- ---- ---- ---- ---- ---- ----
   1.7642338446e+00 Num0 Num0 Num0 Num0 Num0 Num0   P0
   7.8350046724e-04 Num1 Num0 Num0 Num0 Num0 Num0   P0
   4.5430569440e-03 Num0 Num1 Num0 Num0 Num0 Num0   P0
   8.4910279301e-01 Num0 Num0 Num1 Num0 Num0 Num0   P0
  -9.0134378903e-06 Num0 Num0 Num0 Num1 Num0 Num0   P0
   2.6503945379e+00 Num0 Num0 Num0 Num0 Num1 Num0   P0
  -1.3398933210e-02 Num0 Num0 Num0 Num0 Num0 Num1   P0
  -1.1126509255e-01 Num0 Num0 Num0 Num0 Num0 Num0   P1


u1 to u6 are bounded normal variables here, u7 has uniform distribution.

According to the User Manual:

Extended (default if no option is selected): The Extended option avoids the use of any nonlinear variable transformations by augmenting the Askey approach with numerically-generated orthogonal polynomials for non-Askey probability density functions. Extended polynomial selections with numerically-generated polynomials that are orthogonal to the prescribed probability density functions replace each of the sub-optimal Askey basis selections for bounded normal, lognormal, bounded lognormal, loguniform, triangular, gumbel, frechet, weibull, and bin-based histogram.

My old version 6.16.0 Theory Manual gives some additional information:

An alternative is to numerically generate the orthogonal polynomials (using Gauss-Wigert [120], discretized Stieltjes [59], Chebyshev [59], or Gramm-Schmidt [138] approaches) and then compute their Gauss points and weights (using the Golub-Welsch [69] tridiagonal eigensolution). These solutions are optimal for given random variable sets having arbitrary probability density functions and eliminate the need to induce additional nonlinearity through variable transformations, but performing this process for general joint density functions with correlation is a topic of ongoing research (refer to Section 3.5 for additional details).

Where can I find more information on these numerically-generated orthogonal polynomials? Would it be possible for me to reconstruct the gPCE using the coefficients provided by Dakota?

**Comment by briadam** [comment]
In case helpful, when you migrate to a newer version of Dakota, I believe you'll find similar information here (Theory section of the integrated User manual):
 * https://snl-dakota.github.io/docs/6.22.0/users/usingdakota/theory/stochastic.html#askey-scheme
 * https://snl-dakota.github.io/docs/6.22.0/users/usingdakota/theory/stochastic.html#transformations-to-uncorrelated-standard-variables

More generally, this is a guide to map from the old manuals content to the new integrated user manual: https://snl-dakota.github.io/docs/6.22.0/users/classictoc.html
